### PR TITLE
Octagon prototype

### PR DESCRIPTION
**Pull request by a-malyshev**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ members = [
     "examples/tag_analysis/untrustworthy_inputs",
     "examples/tag_analysis/verification_status",
     "examples/tag_analysis/trait_methods",
+    "examples/false-assumptions",
```
Comment (hermanventer) [inline Cargo.toml]: This seems unrelated.

File: checker/src/abstract_value.rs [hunk 2]
```diff
@@ -316,13 +332,16 @@ impl AbstractValue {
                 expression_size: 1,
                 interval: RefCell::new(Some(Rc::new(interval))),
                 tags: RefCell::new(Some(Rc::new(tags))),
+                // todo: add Some value
```
Comment (hermanventer) [inline checker/src/abstract_value.rs]: Presumably you should add octagon from line 325 here.

File: checker/src/abstract_value.rs [hunk 3]
```diff
@@ -1480,6 +1502,7 @@ impl AbstractValueTrait for Rc<AbstractValue> {
         if let Some(result) = self
             .get_cached_interval()
             .greater_or_equal(&other.get_cached_interval())
+            //TODO: add comparison for octagons
```
Comment (hermanventer) [inline checker/src/abstract_value.rs]: You would do that as a separate statement following this one. While the cost/benefit impact is still unclear, it may wise to add an option that controls this.

File: checker/src/abstract_value.rs [hunk 4]
```diff
@@ -2691,6 +2729,60 @@ impl AbstractValueTrait for Rc<AbstractValue> {
...
+    }
+
+    fn get_as_octagon(&self) -> OctagonDomain {
+        if self.expression_size > k_limits::MAX_EXPRESSION_SIZE / 10 {
+            return octagon_domain::BOTTOM;
+        }
+        match &self.expression {
+            Expression::Top => octagon_domain::BOTTOM,
+            Expression::CompileTimeConstant(ConstantDomain::I128(val)) => (*val).into(),
+            //Expression::CompileTimeConstant(ConstantDomain::U128(val)) => (*val).into(),
```
Comment (hermanventer) [inline checker/src/abstract_value.rs]: If val fits into an I128 you could still produce a domain value for this expression.

File: checker/src/octagon_domain.rs [hunk 5]
```diff
@@ -0,0 +1,238 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+//
+
```
Comment (hermanventer) [inline checker/src/octagon_domain.rs:6]: It would be good to add a short description of this intuition behind the Octagon domain here.

**Review by hermanventer — COMMENTED** [review]
Nice to see some action here.

**Comment by a-malyshev** [comment]
Thank you for the comments, will go over them tomorrow

File: checker/src/abstract_value.rs [hunk 6]
```diff
@@ -1948,7 +2002,8 @@ impl AbstractValueTrait for Rc<AbstractValue> {
             }
         };
         let interval = self.get_cached_interval().mul(&other.get_cached_interval());
-        if interval.is_contained_in(&target_type) {
+        let octagon = self.get_cached_octagon().mul(&other.get_cached_octagon());
```
Comment (hermanventer) [inline checker/src/abstract_value.rs:2005]: Only get the octagon if the interval does not answer the question. Likewise for similar situations below.

File: checker/src/abstract_value.rs [hunk 7]
```diff
@@ -2691,6 +2751,62 @@ impl AbstractValueTrait for Rc<AbstractValue> {
...
+    }
+
+    /// Constructs an element of the Octagon domain for simple expressions.
+    #[logfn_inputs(TRACE)]
+    fn get_as_octagon(&self) -> OctagonDomain {
+        if self.expression_size > k_limits::MAX_EXPRESSION_SIZE / 10 {
+            return octagon_domain::BOTTOM;
+        }
+        match &self.expression {
+            Expression::Top => octagon_domain::BOTTOM,
```
Comment (hermanventer) [inline checker/src/abstract_value.rs:2775]: I think this should be TOP.

File: checker/src/abstract_value.rs [hunk 8]
```diff
@@ -2691,6 +2751,62 @@ impl AbstractValueTrait for Rc<AbstractValue> {
...
+                octa
+            }
+            Expression::Sub { left, right } => left.get_as_octagon().sub(&right.get_as_octagon()),
+            Expression::Switch { cases, default, .. } => cases
+                .iter()
+                .fold(default.get_as_octagon(), |acc, (_, result)| {
+                    acc.widen(&result.get_as_octagon())
+                }),
+            _ => {
+                octagon_domain::BOTTOM
```
Comment (hermanventer) [inline checker/src/abstract_value.rs:2803]: This should be BOTTOM only when self is BOTTOM. How are you going to model variables? More specifically, how are you going to deal with expressions involving more variables than the domain can support?

File: checker/src/octagon_domain.rs [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+/// Each cell of the matrix represents potential constraint of the form Vi-Vj<=c,
+/// where c is some constant, Vi and Vj are variables. 
+/// For instance, DBM can be represented as follows 
+///   | V1 |  V2 |  V3 |  V4 |
+/// V1|  0 | -10 |  -5 |  15 |
+/// V2| 10 |   0 | -15 |   5 |
+/// V3| -5 |  15 |   0 | -20 |
+/// V4| 15 |   5 |  20 |   0 |
+///
+/// The matrix has twice more columns and rows because each variable represents
```
Comment (hermanventer) [inline checker/src/octagon_domain.rs:23]: Can you put a reference here?

File: checker/src/abstract_value.rs (hunk 6 above)
Comment (a-malyshev) [reply checker/src/abstract_value.rs:2005]: Will do that. I should have mentioned that I was going to add this later, but forgot that.
Thanks for pointing out it.

File: checker/src/abstract_value.rs (hunk 7 above)
Comment (a-malyshev) [reply checker/src/abstract_value.rs:2775]: oh, sure!

File: checker/src/abstract_value.rs (hunk 8 above)
Comment (a-malyshev) [reply checker/src/abstract_value.rs:2803]: That's a good  question, I don't clearly see how to model it yet. My initial plan was to make the simplest version of the OD and then gradually extend it. Does it sound ok to return to the question once the OD will work well for the simple scenarios?

Comment (hermanventer) [reply checker/src/abstract_value.rs:2803]: Without variables the constant domain will take precedence and nothing interesting will happen in the octal domain. You could, however, get started (locally) by simply assuming there are at most two variables and cut your teeth on test cases that have at most two variables. Eventually, you'll probably have to track which variables are in the domain and go to top when a transfer function attempts to include more than two.

Comment (a-malyshev) [reply checker/src/abstract_value.rs:2803]: I see. It perhaps better to model variables right away. I will think about how to address it.
